### ISSUE TITLE
Cloning Error during pull step in deployment only after first run

### ISSUE THREAD

**Issue by mortengan** [PrefectHQ/prefect#10420]
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

During the pull step in my deployment I clone my private repository from my bitbucket account, to ensure that it always stays up-to-date. The first time, on a fresh install, the run is successful, the second time it fails, with the reason:
```
fatal: destination path 'pr_offer_check' already exists and is not an empty directory.
```
It will also fail all subsequent times until the source directory with the repository is manually deleted.

### Reproduction

```python3
## prefect.yaml
# Generic metadata about this project
name: pr_offer_check
prefect-version: 2.11.0

# build section allows you to manage and build docker images
build: null

# push section allows you to manage if and how this project is uploaded to remote locations
push: null

# pull section allows you to provide instructions for cloning this project in remote locations
pull:
  - prefect.deployments.steps.set_working_directory:
      directory: /mnt

  - prefect.deployments.steps.git_clone:
      repository: https://bitbucket.org/mortendolph/pr_offer_check.git 
      branch: prefect_demo
      access_token: "{{ prefect.blocks.secret.bitbucket-token-pr-offer }}"

  - prefect.deployments.steps.set_working_directory:
      directory: /mnt/pr_offer_check

  - prefect.deployments.steps.pip_install_requirements:
      requirements_file: requirements.txt
      stream_output: True

# the deployments section allows you to provide configuration for deploying flows
deployments:
- name: pr offer check
  version: null
  tags: []
  description: null
  entrypoint: main.py:main
  parameters: {}
  work_pool:
    name: pr_offer_check
    work_queue_name: null
    job_variables: {}
  schedule: null

## run the code twice
prefect deployment run 'main/pr offer check'
prefect deployment run 'main/pr offer check'

## The python code does not matter, as the error occurs before any execution of custom files.
```

### Error

```python3
# LOGS FROM THE DOCKER CONTAINER FROM THE FIRST (SUCCESSFUL) RUN:
06:09:34.710 | INFO    | prefect.flow_runs.worker - Worker 'ProcessWorker 2d0fc7b7-ecaf-451c-b756-83fb5f1314e6' submitting flow run 'd1864706-524b-4b86-9699-68636f5c971d'
06:09:35.329 | INFO    | prefect.flow_runs.worker - Opening process...
06:09:35.484 | INFO    | prefect.flow_runs.worker - Completed submission of flow run 'd1864706-524b-4b86-9699-68636f5c971d'
<frozen runpy>:128: RuntimeWarning: 'prefect.engine' found in sys.modules after import of package 'prefect', but prior to execution of 'prefect.engine'; this may result in unpredictable behaviour
Cloning into 'pr_offer_check'...
06:09:43.435 | INFO    | prefect.deployment - Cloned repository 'https://bitbucket.org/mortendolph/pr_offer_check.git' into 'pr_offer_check'
 > Running set_working_directory step...
 > Running git_clone step...
 > Running set_working_directory step...
 > Running pip_install_requirements step...
Requirement already satisfied: certifi==2023.7.22 in /usr/local/lib/python3.11/site-packages (from -r requirements.txt (line 1)) (2023.7.22)
Requirement already satisfied: charset-normalizer==3.2.0 in /usr/local/lib/python3.11/site-packages (from -r requirements.txt (line 4)) (3.2.0)
Requirement already satisfied: idna==3.4 in /usr/local/lib/python3.11/site-packages (from -r requirements.txt (line 80)) (3.4)
Collecting numpy==1.25.1
  Downloading numpy-1.25.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.6 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.6/17.6 MB 8.1 MB/s eta 0:00:00
Collecting pandas==1.5.3
  Downloading pandas-1.5.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.0 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.0/12.0 MB 10.6 MB/s eta 0:00:00
Requirement already satisfied: python-dateutil==2.8.2 in /usr/local/lib/python3.11/site-packages (from -r requirements.txt (line 137)) (2.8.2)
Requirement already satisfied: pytz==2023.3 in /usr/local/lib/python3.11/site-packages (from -r requirements.txt (line 140)) (2023.3)
Requirement already satisfied: requests==2.31.0 in /usr/local/lib/python3.11/site-packages (from -r requirements.txt (line 143)) (2.31.0)
Requirement already satisfied: six==1.16.0 in /usr/local/lib/python3.11/site-packages (from -r requirements.txt (line 146)) (1.16.0)
Requirement already satisfied: urllib3==1.26.16 in /usr/local/lib/python3.11/site-packages (from -r requirements.txt (line 149)) (1.26.16)
Installing collected packages: numpy, pandas
Successfully installed numpy-1.25.1 pandas-1.5.3
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
[notice] A new release of pip available: 22.3.1 -> 23.2.1
[notice] To update, run: pip install --upgrade pip
06:10:13.247 | INFO    | Flow run 'talented-lynx' - Finished in state Completed()
06:10:14.534 | INFO    | prefect.flow_runs.worker - Process 14 exited cleanly.

# LOGS FROM THE DOCKER CONTAINER FROM THE SECOND (FAILED) RUN:
06:10:25.307 | INFO    | prefect.flow_runs.worker - Worker 'ProcessWorker 2d0fc7b7-ecaf-451c-b756-83fb5f1314e6' submitting flow run '5f362017-98c6-44fb-946e-9bb8983ffb94'
06:10:25.937 | INFO    | prefect.flow_runs.worker - Opening process...
06:10:26.129 | INFO    | prefect.flow_runs.worker - Completed submission of flow run '5f362017-98c6-44fb-946e-9bb8983ffb94'
<frozen runpy>:128: RuntimeWarning: 'prefect.engine' found in sys.modules after import of package 'prefect', but prior to execution of 'prefect.engine'; this may result in unpredictable behaviour
fatal: destination path 'pr_offer_check' already exists and is not an empty directory.
06:10:31.819 | ERROR   | Flow run 'nondescript-aardwark' - Flow could not be retrieved from deployment.
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/prefect/deployments/steps/core.py", line 124, in run_steps
    step_output = await run_step(step, upstream_outputs)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/prefect/deployments/steps/core.py", line 95, in run_step
    result = await from_async.call_soon_in_new_thread(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/prefect/_internal/concurrency/calls.py", line 292, in aresult
    return await asyncio.wrap_future(self.future)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/prefect/_internal/concurrency/calls.py", line 316, in _run_sync
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/prefect/deployments/steps/pull.py", line 144, in git_clone
    raise RuntimeError(
RuntimeError: Failed to clone repository 'https://bitbucket.org/mortendolph/pr_offer_check.git' with exit code 128.
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/prefect/engine.py", line 395, in retrieve_flow_then_begin_flow_run
    flow = await load_flow_from_flow_run(flow_run, client=client)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/prefect/client/utilities.py", line 51, in with_injected_client
    return await fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/prefect/deployments/deployments.py", line 207, in load_flow_from_flow_run
    output = await run_steps(deployment.pull_steps)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/prefect/deployments/steps/core.py", line 152, in run_steps
    raise StepExecutionError(f"Encountered error while running {fqn}") from exc
prefect.deployments.steps.core.StepExecutionError: Encountered error while running prefect.deployments.steps.git_clone
 > Running set_working_directory step...
 > Running git_clone step...
06:10:33.300 | INFO    | prefect.flow_runs.worker - Process 47 exited cleanly.

# OUTPUT LOGS FROM PREFECT CLOUD WEBSITE (could not find a way to paste them in a nice format):
Worker 'ProcessWorker 2d0fc7b7-ecaf-451c-b756-83fb5f1314e6' submitting flow run '5f362017-98c6-44fb-946e-9bb8983ffb94'
08:10:25 AM
prefect.flow_runs.worker

Opening process...
08:10:25 AM
prefect.flow_runs.worker

Completed submission of flow run '5f362017-98c6-44fb-946e-9bb8983ffb94'
08:10:26 AM
prefect.flow_runs.worker

Flow could not be retrieved from deployment.
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/prefect/deployments/steps/core.py", line 124, in run_steps
    step_output = await run_step(step, upstream_outputs)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/prefect/deployments/steps/core.py", line 95, in run_step
    result = await from_async.call_soon_in_new_thread(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/prefect/_internal/concurrency/calls.py", line 292, in aresult
    return await asyncio.wrap_future(self.future)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/prefect/_internal/concurrency/calls.py", line 316, in _run_sync
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/prefect/deployments/steps/pull.py", line 144, in git_clone
    raise RuntimeError(
RuntimeError: Failed to clone repository 'https://bitbucket.org/mortendolph/pr_offer_check.git' with exit code 128.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/prefect/engine.py", line 395, in retrieve_flow_then_begin_flow_run
    flow = await load_flow_from_flow_run(flow_run, client=client)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/prefect/client/utilities.py", line 51, in with_injected_client
    return await fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/prefect/deployments/deployments.py", line 207, in load_flow_from_flow_run
    output = await run_steps(deployment.pull_steps)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/prefect/deployments/steps/core.py", line 152, in run_steps
    raise StepExecutionError(f"Encountered error while running {fqn}") from exc
prefect.deployments.steps.core.StepExecutionError: Encountered error while running prefect.deployments.steps.git_clone
08:10:31 AM
prefect.flow_runs

Process 47 exited cleanly.
08:10:33 AM
prefect.flow_runs.worker
```

### Versions

```Text
# Work environment on my laptop
Version:             2.11.0
API version:         0.8.4
Python version:      3.10.12
Git commit:          eeb9e219
Built:               Thu, Jul 20, 2023 4:34 PM
OS/Arch:             linux/x86_64
Profile:             default
Server type:         cloud

# The version running the worker is given in the docker image.
```

### Additional context

**Setup**
I have a linux server, which runs a docker image that automatically starts a worker and connects to a work pool. The docker image is running 24/7. My script is backed up on a private bitbucket repository and I use Prefect Cloud to keep track of everything. My `prefect.yaml` files includes a pull step, where it will clone the repository and keep it up-to-date from that point on. The `prefect.yaml` file looks like this:

```
# Generic metadata about this project
name: pr_offer_check
prefect-version: 2.11.0

# build section allows you to manage and build docker images
build: null

# push section allows you to manage if and how this project is uploaded to remote locations
push: null

# pull section allows you to provide instructions for cloning this project in remote locations
pull:
  - prefect.deployments.steps.set_working_directory:
      directory: /mnt

  - prefect.deployments.steps.git_clone:
      repository: https://bitbucket.org/mortendolph/pr_offer_check.git 
      branch: prefect_demo
      access_token: "{{ prefect.blocks.secret.bitbucket-token-pr-offer }}"

  - prefect.deployments.steps.set_working_directory:
      directory: /mnt/pr_offer_check

  - prefect.deployments.steps.pip_install_requirements:
      requirements_file: requirements.txt
      stream_output: True

# the deployments section allows you to provide configuration for deploying flows
deployments:
- name: pr offer check
  version: null
  tags: []
  description: null
  entrypoint: main.py:main
  parameters: {}
  work_pool:
    name: pr_offer_check
    work_queue_name: null
    job_variables: {}
  schedule: null
```

The `Dockerfile` and `compose.yaml` file to create the docker image and start the worker is copied below

_Dockerfile_
```
# Install he prefect image and login.
FROM prefecthq/prefect:2-python3.11 AS builder

# Cant start worker from first stage, as CMD will override all CMD's in the original prefect image.
# Therefore, a second stage is added, where a CMD can be freely added without consequences for previous
# CMD's.
FROM builder AS starter

# RUN pip install poetry

CMD prefect worker start --pool pr_offer_check
```

_compose.yaml_
```
services:
  worker:
    build:
      context: .
    env_file:
      - ../.env
    volumes:
      - /home/morten/Projects/ToSend:/mnt/ToSend
    container_name: prefect-worker-pr-offer-check
    restart: unless-stopped
```

**Comment by desertaxle** [comment]
Thanks for the issue @mortengan! We need to update the `git_clone` step to correctly handle situations where the repository has already been cloned or a file with the same name already exists.

In the meantime, we might be able to figure out a workaround. Is there a dependency that your flow needs in the `/mnt` directory? Changing the directory to `/mnt` overrides the worker's default behavior of cloning into a short-lived directory, so if we can remove that part we might be able to craft a `pull` deploy action that will work consistently.

**Comment by desertaxle** [comment]
`GitRepository.pull_code()` (which underpins the `git_clone` step) now handles pre-existing repos with `git pull`/`git fetch` and falls back to `shutil.rmtree()` + re-clone on errors.

If you run into any other issues with `git` cloning, please open a new issue!
